build.gradle

```diff
@@ -102,25 +120,61 @@ dependencies {
     // These dependencies are only needed for running tests
 
     testImplementation 'org.hamcrest:hamcrest:2.2'
-    testImplementation "org.junit.jupiter:junit-jupiter-api:5.6.2"
-    testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine:5.6.2"
+    testImplementation "org.junit.jupiter:junit-jupiter-api:5.6.+"
+    testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine:5.6.+"
 
-    testImplementation group: 'org.mockito', name: 'mockito-inline', version: '3.3.3' // For final classes
+    testImplementation('org.mockito:mockito-inline:3.3.+') {
+        because "-inline build enables mocking final classes"
+        // https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#0.2
+        // > Be aware that this artifact may be abolished when the inline mock making feature is integrated into the default mock maker.
+    }
+    testImplementation "org.mockito:mockito-junit-jupiter:3.3.+"
+
+    testImplementation('org.spf4j:spf4j-slf4j-test:8.8.+') {
+        because "testable logging"
+    }
+    testImplementation("org.slf4j:slf4j-api:1.7.+!!") {
```

Comment:
`.+!!"` the elder speak to me - what do they want to tell me o.O

Reply:
`1.7.+` [prefixed version range](https://docs.gradle.org/current/userguide/single_versions.html), any 1.7.x version

`!!` [strict version](https://docs.gradle.org/current/userguide/rich_versions.html). that is, if something else pulled in slf4j-api 1.8, do _not_ consider that to satisfy this dependency, because it does not match 1.7.

oh, that page shows there are ways we could write this out in the DSL instead of the symbol-heavy shorthand:

```kotlin
testImplementation("org.slf4j:slf4j-api") {
    version {
        strictly("1.7.+")
    }
}
```